mongorm/BaseDocument.py

```diff
@@ -7,31 +7,35 @@
 
 class BaseDocument(with_metaclass(DocumentMetaclass, object)):
 	__internal__ = True
-	
+
 	class DoesNotExist(Exception):
 		pass
-	
+
 	def __init__( self, **kwargs ):
 		self._is_lazy = False
 		self._data = {}
 		self._values = {}
-		
+
 		self._data['_types'] = serialiseTypesForDocumentType( self.__class__ )
-		
+
 		for name,value in kwargs.items( ):
 			setattr(self, name, value)
-	
+
 	def _fromMongo( self, data, overwrite=True ):
 		self._is_lazy = True
-		
+
 		for (name,field) in self._fields.items( ):
 			dbField = field.dbField
 			if dbField in data and ( overwrite or not name in self._values ):
 				pythonValue = field.toPython( data[dbField] )
 				setattr(self, name, pythonValue)
-		
+
 		return self
 
+	# For python 2/3 compatibility
+	def __bool__( self ):
+		return True
```

Comment:
objects are not truthy by default?

Reply:
![image](https://user-images.githubusercontent.com/5072366/68717849-e9765d00-05fb-11ea-8990-f2b095840336.png)

Neither in python 2 or 3 are empty objects considered `truthy`. But because the file overwrites `__getattr__` (so that it returns `None` instead of an `AttributeError`), the way the `future` `object` class checks truthiness makes it error because it believes the attribute (the `__bool__` function) exists on itself, but when it calls it, it dies. But the old behaviour of the library is that an empty Document is truthy, so this just keeps that compatibility

Reply:
damn

Reply:
> ![image](https://user-images.githubusercontent.com/5072366/68717849-e9765d00-05fb-11ea-8990-f2b095840336.png)
> 
> Neither in python 2 or 3 are empty objects considered `truthy`. But because the file overwrites `__getattr__` (so that it returns `None` instead of an `AttributeError`), the way the `future` `object` class checks truthiness makes it error because it believes the attribute (the `__bool__` function) exists on itself, but when it calls it, it dies. But the old behaviour of the library is that an empty Document is truthy, so this just keeps that compatibility

That sounds like a bug; `__getattr__` should raise an `AttributeError` for things that aren't fields. Actually all objects in Python are considered `True` unless they specify a `__bool__` method which returns `False` or  `__len__` method which returns 0. Also if you decide to go ahead with this workaround you should probably also specify a `__nonzero__` method for Python 2.x (`__bool__` was added in Python 3.x)

Reply:
Yeah, I don't like that `__getattr__` doesn't raise an `AttributeError` either ahaha. But it's been that way in the library since it's inception it seems, so I don't want to break functionality as core as that in OpenLearning which expects it. After a better inspection across the whole OpenLearning codebase to check it's uses, I'd be more comfortable to recommend that we make that change!

> Actually all objects in Python are considered `True`

Yeap, whoops. Working on JS my brain confused object and dict. Yeap, it's not that `object` is `falsy`, it's that the `future` version of `object` that we're importing (aka ```from builtins import object```) [first checks if the `__bool__` method is available and uses that to determine it's truthiness (because it defines the `__nonzero__` method) that python 2 uses](https://github.com/PythonCharmers/python-future/blob/master/src/future/types/newobject.py#L68-L74). But because of the above `AttributeError` not being raised, it tries to call that method, and dies because it doesn't actually exist to be called. So this was just there to short circuit it into an acceptable response that we want!